composite/token-allocation/src/adapter.ts

```diff
@@ -35,14 +36,36 @@ export const marketCapTotalValue = (
     .toNumber()
 }
 
-const toValidAllocations = (allocations: TokenAllocations): TokenAllocations => {
-  if (!allocations.every((t) => !!t.symbol))
-    throw new AdapterError({ message: `Symbol not available for all tokens.`, statusCode: 400 })
-  return allocations.map((t) => ({
-    symbol: t.symbol.toUpperCase(),
-    decimals: t.decimals || DEFAULT_TOKEN_DECIMALS,
-    balance: t.balance || DEFAULT_TOKEN_BALANCE * 10 ** (t.decimals || DEFAULT_TOKEN_DECIMALS),
-  }))
+const toValidAllocations = (allocations: any[]): TokenAllocations => {
+  const _toValidSymbol = (symbol: string) => {
+    if (!symbol)
+      throw new AdapterError({ message: `Symbol not available for all tokens.`, statusCode: 400 })
+    return symbol.toUpperCase()
+  }
+  const _toValidDecimals = (decimals: number | undefined) => {
+    if (decimals === undefined) return DEFAULT_TOKEN_DECIMALS
+    return Number.isInteger(decimals) && decimals >= 0 ? decimals : DEFAULT_TOKEN_DECIMALS
+  }
+  const _toValidBalance = (balance: number | string | undefined, decimals: number) => {
+    if (!balance) return DEFAULT_TOKEN_BALANCE * 10 ** decimals
+    let BNbalance
+    try {
+      BNbalance = BigNumber.from(balance)
+    } catch (e) {
+      throw new AdapterError({ message: `Invalid balance: ${e.message}`, statusCode: 400 })
+    }
+    if (!BNbalance.gte(0))
```

Comment:
We can make this easier to read

```suggestion
    if (BNbalance.lt(0))
```

Reply:
Or we could do `.isNegative()` ?